src/main/scala/org/ergoplatform/nodeView/history/extra/ExtraIndexer.scala

```diff
@@ -114,73 +100,74 @@ trait ExtraIndexerBase extends ScorexLogging {
-      if (trees(i).id == id) { // address found in last saveLimit modifiers
-        spendOrReceive match {
-          case Left(box) => trees(i).addTx(globalTxIndex).spendBox(box) // spend box
-          case Right(iEb) => trees(i).addTx(globalTxIndex).addBox(iEb) // receive box
-        }
-        return i
+  private def findAndUpdateTree(id: ModifierId, spendOrReceive: Either[IndexedErgoBox, IndexedErgoBox]): Unit = {
+    trees.get(id).map(tree => {
+      spendOrReceive match {
+        case Left(iEb) => tree.addTx(globalTxIndex).spendBox(iEb, Some(history)) // spend box
+        case Right(iEb) => tree.addTx(globalTxIndex).addBox(iEb) // receive box
       }
-    }
+      return
+    })
     history.typedExtraIndexById[IndexedErgoAddress](id) match { // address not found in last saveLimit modifiers
       case Some(x) =>
         spendOrReceive match {
-          case Left(box) => trees += x.addTx(globalTxIndex).spendBox(box) // spend box
-          case Right(iEb) => trees += x.addTx(globalTxIndex).addBox(iEb) // receive box
+          case Left(box) => trees.put(id, x.addTx(globalTxIndex).spendBox(box, Some(history))) // spend box
+          case Right(iEb) => trees.put(id, x.addTx(globalTxIndex).addBox(iEb)) // receive box
         }
-        trees.length - 1
       case None => // address not found at all
         spendOrReceive match {
-          case Left(box) => log.warn(s"Unknown address spent box ${bytesToId(box.id)}") // spend box should never happen by an unknown address
-          case Right(iEb) => trees += IndexedErgoAddress(id, ListBuffer(globalTxIndex), ListBuffer.empty[Long], Some(new BalanceInfo)).addBox(iEb) // receive box
+          case Left(iEb) => log.warn(s"Unknown address spent box ${bytesToId(iEb.box.id)}") // spend box should never happen by an unknown address
+          case Right(iEb) => trees.put(id, IndexedErgoAddress(id, ArrayBuffer(globalTxIndex), ArrayBuffer.empty[Long], Some(BalanceInfo())).addBox(iEb)) // receive box
         }
-        -1
     }
   }
 
   /**
     * @return number of indexes in all buffers
     */
-  private def modCount: Int = general.length + boxes.length + trees.length
+  private def modCount: Int = general.length + boxes.size + trees.size
 
   /**
     * Write buffered indexes to database and clear buffers.
     */
   private def saveProgress(writeLog: Boolean = true): Unit = {
 
-    if(modCount == 0) return
-
-    val start: Long = System.nanoTime()
+    val start: Long = System.currentTimeMillis
 
-    // perform segmentation on big modifiers
-    val addressesLen: Int = trees.length
-    cfor(0)(_ < addressesLen, _ + 1) { i =>
-      if (trees(i).txs.length > segmentTreshold || trees(i).boxes.length > segmentTreshold) trees ++= trees(i).splitToSegments()
+    // perform segmentation on big addresses and save their internal segment buffer
+    trees.values.foreach { tree =>
+      if(tree.segments.nonEmpty) {
+        tree.segments.foreach(seg => trees.put(seg.id, seg))
```

Comment:
Took a look at the implementation and it seems safe

Reply:
@jellymlg, does it iterate over newly added values? Or it first copies values to a separate collection and then iterates over it?

Reply:
```
/** Avoid iterator for a 2x faster traversal. */
  protected def foreachEntry[U](f: Entry => U) {
    val iterTable = table
    var idx       = lastPopulatedIndex
    var es        = iterTable(idx)

    while (es != null) {
      val next = es.next // Cache next in case f removes es.
      f(es.asInstanceOf[Entry])
      es = next

      while (es == null && idx > 0) {
        idx -= 1
        es = iterTable(idx)
      }
    }
  }```

Reply:
Made a different buffer for segments.